### PR TITLE
Files from remote machine

### PR DESCRIPTION
**Pull request by jonahfried**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -27,3 +27,7 @@ From classifier.py -h:
 > -                     [--num_train_examples NUM_TRAIN_EXAMPLES]
 > -                     [--num_test_examples NUM_TEST_EXAMPLES]
                      
+
+# document_vector.ipynb
+Visualizes the spread of data and their similarities to one another. 
+use extract_features.py to prepare json for document_vector notebook.
```
Comment (thatandromeda) [inline README.md:33]: I find it handy to have commands in copy/pastable form, because sometimes English-language descriptions leave out flags or something and you spend a long time rederiving command syntax. So something like, "To prepare json for document_vector notebook: `extract_features.py your_file_name.json`." (or whatever the syntax actually is)

File: classifier.py [hunk 2]
```diff
@@ -0,0 +1,261 @@
...
+from tqdm import tqdm, trange
+
+from pytorch_pretrained_bert.file_utils import PYTORCH_PRETRAINED_BERT_CACHE
+from pytorch_pretrained_bert.modeling import BertForSequenceClassification, BertConfig, WEIGHTS_NAME, CONFIG_NAME
+from pytorch_pretrained_bert.tokenization import BertTokenizer
+from pytorch_pretrained_bert.optimization import BertAdam, warmup_linear
+
+from data_processing import *
+from classifier_parser import get_args
+from classifier_logging import *
```
Comment (thatandromeda) [inline classifier.py:18]: Nice organization on the imports.

File: classifier.py [hunk 3]
```diff
@@ -0,0 +1,261 @@
...
+    """ Saves the model to the given output directory """
+    # Save a trained model and the associated configuration
+    model_to_save = model.module if hasattr(model, 'module') else model  # Only save the model it-self
+    output_model_file = os.path.join(output_dir, WEIGHTS_NAME)
+    torch.save(model_to_save.state_dict(), output_model_file)
+    output_config_file = os.path.join(output_dir, CONFIG_NAME)
+    with open(output_config_file, 'w') as f:
+        f.write(model_to_save.config.to_json_string())
+
+def get_device_and_n_gpu(use_cpu, local_rank):
```
Comment (thatandromeda) [inline classifier.py:233]: If it's possible to separate this into `get_device` and `get_n_gpu`, that would be preferable (makes it much easier in the future to change how you get either of these without needing to disentangle it from the other).

File: classifier.py [hunk 4]
```diff
@@ -0,0 +1,261 @@
...
+from pytorch_pretrained_bert.file_utils import PYTORCH_PRETRAINED_BERT_CACHE
+from pytorch_pretrained_bert.modeling import BertForSequenceClassification, BertConfig, WEIGHTS_NAME, CONFIG_NAME
+from pytorch_pretrained_bert.tokenization import BertTokenizer
+from pytorch_pretrained_bert.optimization import BertAdam, warmup_linear
+
+from data_processing import *
+from classifier_parser import get_args
+from classifier_logging import *
+
+def main():
```
Comment (thatandromeda) [inline classifier.py:20]: There is a lot of stuff going on in main() and it would be helpful to have a comment above every chunk of functionality saying what's going on. (Or to pull out cleanly separable chunks into functions with clear names; the names then serve the same purpose as the comments would have.)

File: classifier.py [hunk 5]
```diff
@@ -0,0 +1,261 @@
...
+    num_labels = args.num_labels
+
+    create_output_dir(args.output_dir, args.force)
+    processor.check_data_exists(args.skip_train)
+
+    model = get_model(args.cache_dir, args.local_rank, args.bert_model, num_labels, device, n_gpu)
+
+    step_tracker = StepValues()
+
+    if not args.skip_train:
```
Comment (thatandromeda) [inline classifier.py:47]: There are so many things passing args around here that I wonder if what we secretly have is some kind of like...Classifier object....that uses args to initalize itself and then has various functions, like `get_device` and `setup_train` and `perform_train`. I'm not saying you should refactor into that (a lot of work for very little benefit at this point, unless you're planning to use this codebase a lot in the future) but you might find it illuminating to watch some of Sandi Metz's talks on software design, e.g. https://www.youtube.com/watch?v=8bZh5LMaSmE (it's in ruby but you can ignore that fact; the language is incidental to the ideas).

File: classifier.py [hunk 6]
```diff
@@ -0,0 +1,261 @@
...
+
+    create_output_dir(args.output_dir, args.force)
+    processor.check_data_exists(args.skip_train)
+
+    model = get_model(args.cache_dir, args.local_rank, args.bert_model, num_labels, device, n_gpu)
+
+    step_tracker = StepValues()
+
+    if not args.skip_train: 
+        step_tracker = setup_and_perform_train(
```
Comment (thatandromeda) [inline classifier.py:48]: `setup_and_perform_train` worries me - I want it to be `setup_train` and `perform_train`. If there's some reason it can't be that, I would wonder if either 1) this is the correct function name or 2) it needs to be rewritten in such a way that it CAN be separated.

File: classifier.py [hunk 7]
```diff
@@ -0,0 +1,261 @@
...
+            args.learning_rate,
+            args.warmup_proportion, 
+            device, 
+            n_gpu
+        )
+        save_model(model, args.output_dir, num_labels)
+    
+
+    if (args.local_rank == -1 or torch.distributed.get_rank() == 0) and not args.skip_eval:
+        setup_and_perform_eval(model, args.num_test_examples, processor, args.output_dir, device, step_tracker)
```
Comment (thatandromeda) [inline classifier.py:63]: Again making me wonder if there is secretly a `setup_eval` and `perform_eval` function here. (Also I would tend away from "eval" if possible, since `eval()` is a python function (and a rather dramatic one at that) - can you find a more descriptive name for what you're doing? I'm not 100% sure what you're evaluating here yet but examples might be `setup_error_function` and `evaluate_error`.

File: classifier.py [hunk 8]
```diff
@@ -0,0 +1,261 @@
...
+
+def setup_and_perform_train( processor, model, num_train_examples, gradient_accumulation_steps,
+    num_train_epochs, local_rank, learning_rate, warmup_proportion, device, n_gpu ):
+    """ Prepares dataloader and optimizer then performs training on the model """
```
Comment (thatandromeda) [inline classifier.py:67]: Python strongly prefers verb forms like "prepare" over "prepares" in a docstring: https://www.python.org/dev/peps/pep-0257/#one-line-docstrings . Also it's spooky how much it improves your comments if you require yourself to end them all with periods (or at least it's spooky how much it improves mine), in addition to being specified by PEP 257.

File: classifier.py [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+    train_dataloader = processor.get_train_dataloader(local_rank, num_train_examples)
+
+    num_train_optimization_steps = processor.get_num_train_optimization_steps(gradient_accumulation_steps, num_train_epochs, local_rank)
+    log_training(num_train_examples, processor.train_batch_size, num_train_optimization_steps)
+    
+    optimizer = get_optimizer(model, learning_rate, warmup_proportion, num_train_optimization_steps)
```
Comment (thatandromeda) [inline classifier.py:73]: These are all super good variable names; I know exactly what they're for.

File: classifier.py [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+    step_tracker = train_model(model, optimizer, train_dataloader, num_train_epochs, gradient_accumulation_steps, device, n_gpu)
+    
+    return step_tracker
+    
+
+def train_model(model, optimizer, train_dataloader, num_train_epochs, gradient_accumulation_steps, device, n_gpu):
+    """ Trains the model on each example in the example set. """
+    model.train()
+    global_step = 0
+    for _ in trange(int(num_train_epochs), desc="Epoch"):
```
Comment (thatandromeda) [inline classifier.py:84]: This for loop definitely wants a comment, and possibly some additional comments inside - it's doing a lot.

File: classifier.py [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+def setup_and_perform_eval(model, num_test_examples, processor, output_dir, device, step_tracker):
+    eval_dataloader = processor.get_eval_dataloader(num_test_examples)
+    log_evaluating(processor.num_eval_examples, processor.eval_batch_size)
+    
+    result = eval_model(model, device, eval_dataloader, step_tracker)
+    
+    log_result(output_dir, processor.num_train_examples, processor.num_eval_examples, result)
+    record_to_csv(output_dir, processor.num_train_examples, processor.num_eval_examples, result)
+
+def eval_model(model, device, eval_dataloader, step_tracker):
```
Comment (thatandromeda) [inline classifier.py:138]: Also a function that's doing a lot of work, ergo one where inline comments (or extraction into smaller functions) would be a kindness to readers.

File: classifier.py [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+        eval_loss += tmp_eval_loss.mean().item()
+        eval_accuracy += tmp_eval_accuracy
+
+        nb_eval_examples += input_ids.size(0)
+        nb_eval_steps += 1
+
+    eval_loss = eval_loss / nb_eval_steps
+    eval_accuracy = eval_accuracy / nb_eval_examples
+    loss = step_tracker.loss()
+    result = {
```
Comment (thatandromeda) [inline classifier.py:168]: This dict gets passed around so much I wonder if it's possible it wants to be an object, with these values as attributes. That would let you perform validations on  it, have functions that attach specifically to the data (e.g. a `Result.write_to_csv` that understands how results should be formatted), and work with it more easily in test suites. (Recognizing that testing ML code is a nightmare.)

File: classifier.py [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+    return result
+
+def accuracy(out, labels):
+    """ returns the number of predictions that equal the correct label """
+    outputs = np.argmax(out, axis=1)
+    return np.sum(outputs == labels)
+
+
+def get_model(cache_dir, local_rank, bert_model, num_labels, device, n_gpu):
+    "Returns a model with the given specifications"
```
Comment (thatandromeda) [inline classifier.py:183]: I don't think this comment is doing enough work as it doesn't say anything I didn't know from the function signature. It should either provide more insight or be dropped.

File: classifier.py [hunk 14]
```diff
@@ -0,0 +1,261 @@
...
+        cache_dir=cache_path,
+        num_labels = num_labels
+    )
+
+    model.to(device)
+    if local_rank != -1:
+        try:
+            from apex.parallel import DistributedDataParallel as DDP
+        except ImportError:
+            raise ImportError("Please install apex from https://www.github.com/nvidia/apex to use distributed and fp16 training.")
```
Comment (thatandromeda) [inline classifier.py:198]: Oooh, that's very helpful.

File: classifier_logging.py [hunk 15]
```diff
@@ -0,0 +1,50 @@
...
+logging.basicConfig(
+    format = '%(asctime)s - %(levelname)s - %(name)s -   %(message)s',
+    datefmt = '%m/%d/%Y %H:%M:%S',
+    level = logging.INFO
+)
+
+logger = logging.getLogger(__name__)
+
+
+def log_training(num_examples, train_batch_size, num_train_optimization_steps):
```
Comment (thatandromeda) [inline classifier_logging.py:15]: Aw yeah, standardized logging statements for the win.

File: classifier_logging.py [hunk 16]
```diff
@@ -0,0 +1,50 @@
...
+            logger.info("  %s = %s", key, str(result[key]))
+            writer.write("%s = %s\n" % (key, str(result[key])))
+
+def record_to_csv(output_dir, num_train_examples, num_test_examples, result):
+    """ creates (if necessary) and stores results in a 'combined.csv' file """
+    path = os.path.join(output_dir, "combined.csv")
+    result["num_train_examples"] = num_train_examples
+    result["num_test_examples"] = num_test_examples
+    results = pd.Series(result)
+    past_results = pd.DataFrame()
```
Comment (thatandromeda) [inline classifier_logging.py:46]: This part might be clearer as
```
if os.path.exists(path):
    past_results = pd.read_csv(path)
else:
    past_results = pd.DataFrame()
```

(just to make it super obvious that `past_results` always exists,  and therefore we will be able to append to it in line 49.)

What happens if `pd.read_csv` throws an exception? (maybe the path is a directory, maybe the file isn't a valid csv file, et cetera)? Do you want to let that bubble up and crash the program (this can be the right call if there's no obvious way for the computer to handle it, or if you just don't know what to expect so you want things to break dramatically if they break so that you will notice)? Do you want to write the current results to CSV and just ignore the past results? Do you want to not bother writing to CSV, display some kind of error message, and let the rest of the program keep executing? (There is no particular right answer here; just that the code should reflect what you pick.)

File: data_processing.py [hunk 17]
```diff
@@ -0,0 +1,128 @@
+import os
+import pandas as pd
+import torch
+from torch.utils.data import (DataLoader, RandomSampler, SequentialSampler, TensorDataset)
+from collections import namedtuple
+
+class LeftRightProcessor:
```
Comment (thatandromeda) [inline data_processing.py:7]: Why does it have this name? Is there another name that would make its function clearer?

File: data_processing.py [hunk 18]
```diff
@@ -0,0 +1,128 @@
...
+    """ The LeftRight Processor handles getting data from csv files to a format useful to the model. """
+    def __init__(self, max_seq_length, tokenizer, data_dir, seed, train_batch_size, eval_batch_size):
+        self.max_seq_length = max_seq_length
+        self.tokenizer = tokenizer
+        self.data_dir = data_dir
+        self.seed = seed
+        self.train_batch_size = train_batch_size
+        self.eval_batch_size = eval_batch_size
+        
+    def get_labels(self):
```
Comment (thatandromeda) [inline data_processing.py:17]: Possibly this wants to be an attribute and not a function (self.labels = `["left", "right"]` in `__init__` e.g.). But if you anticipate the labels changing a lot, or involving conditional logic, a function is a fine place.

File: data_processing.py [hunk 19]
```diff
@@ -0,0 +1,128 @@
...
+        "classification labels for the dataset"
+        return ["left", "right"]
+
+    def check_data_exists(self, skip_train):
+        """ Raise error if can't data can not be found """
+        cant_find_train_data = (not skip_train) and (not os.path.exists(self.data_dir+"/train.csv"))
```
Comment (thatandromeda) [inline data_processing.py:23]: `os.path.join` might be better than the + here

File: extract_features.py [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+    eval_dataset = get_dataset(eval_features)
+    eval_sampler = SequentialSampler(eval_dataset)
+    return DataLoader(eval_dataset, sampler=eval_sampler, batch_size=eval_batch_size)
+
+def _convert_examples_to_features(examples, label_list, max_seq_length, tokenizer):
+    """ convert each example in a list to an InputFeatures """
+    InputFeatures = namedtuple("InputFeatures", ["input_ids", "input_mask", "segment_ids", "label_id", "tokens", "case_id"])
+    
+    labels_to_int = {label:i for i,label in enumerate(label_list)}
+    print(labels_to_int)
```
Comment (thatandromeda) [inline extract_features.py:37]: Is this a debugging statement that should have been removed?

File: extract_features.py [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+    parser.add_argument("--device", type=str, default="cuda",
+        help="where to perform calculations. 'cpu' or 'cuda'"
+    )
+    parser.add_argument("--data_dir", type=str, required=True,
+        help="location of stored data. Must have a test.csv"
+    )
+    return parser.parse_args()
+    
+
+if __name__ == "__main__":
```
Comment (thatandromeda) [inline extract_features.py:103]: Also a very long function that would benefit from inline comments and/or being extracted into smaller functions. (I might simply not bother with extracting variables from `args` and instead just pass `args` around and let functions get what they needed from it, and then have the rest of this section read something like
```
build_model()
predict_labels()
build_feature_dict()
model.eval()
label_predictor.eval()
with open(output_file, "w", encoding='utf-8') as writer:
    do_something()
    do_something_else()
```

(I'm not clear enough on what's happening in the for loop to know what those last functions should be called; also obviously some of these might need to return stuff and pass data, but you get the overall point -- writing it this way would let me read what's happening in this function in outline form rather than essay form, and then I could dip down into the functions for details as needed.)

**Review by thatandromeda — APPROVED** [review]
Ok, I've added my thoughts on software engineering (except for the ipython stuff, which I can't see on github and don't have set up locally). Feel free to change things or not change things as you prefer on this basis :) And then merge when ready.
